solidity/contracts/converter/BancorConverter.sol

```diff
@@ -809,27 +826,29 @@ contract BancorConverter is IBancorConverter, SmartTokenController, ContractRegi
 
         // iterate through the reserve tokens and transfer a percentage equal to the ratio between
         // _amount and the total supply in each reserve from the caller to the converter
-        for (uint16 i = 0; i < reserveTokens.length; i++) {
+        for (uint256 i = 0; i < reserveTokens.length; i++) {
             IERC20Token reserveToken = reserveTokens[i];
-            bool isETH = reserveToken == IERC20Token(0);
-            uint256 reserveBalance = getReserveBalance(reserveToken) - (isETH ? msg.value : 0);
+            uint256 reserveBalance = getReserveBalance(reserveToken);
             uint256 reserveAmount = formula.calculateFundCost(supply, reserveBalance, totalReserveRatio, _amount);
 
             // transfer funds from the caller in the reserve token
-            if (isETH) {
+            if (reserveToken == IERC20Token(0)) {
                 if (msg.value > reserveAmount) {
                     msg.sender.transfer(msg.value - reserveAmount);
                 }
                 else if (msg.value < reserveAmount) {
                     require(msg.value == 0);
-                    ensureTransferFrom(etherToken, msg.sender, this, reserveAmount);
+                    safeTransferFrom(etherToken, msg.sender, this, reserveAmount);
                     etherToken.withdraw(reserveAmount);
                 }
             }
             else {
-                ensureTransferFrom(reserveToken, msg.sender, this, reserveAmount);
+                safeTransferFrom(reserveToken, msg.sender, this, reserveAmount);
             }
 
+            // sync the reserve balance	
+            syncReserveBalance(reserveToken);

```

Comment:
May as well embed the code of function `syncReserveBalance` into the if/else clause above.